src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+    if (!originalString || originalString.length === 0) {
+        return [];
+    }
+    // There is only one "word" in this string.
+    else if (currentPhraseDef === null) {
+        return [{
+            value: originalString,
+            index: 0
+        }];
+    }
+    else {
+        var segments = [];
+        // Check for leading whitespace first.
+        var leadingWhitespaceMatches = originalString.match(/^(\W+)/);
+        if (leadingWhitespaceMatches) {
+            segments.push({
+                value: leadingWhitespaceMatches[1],
+                index: 0
+            });
+        }
+        while (currentPhraseDef !== null) {
+            // The first capture match is a "word"
+            var word = currentPhraseDef[1];
+            segments.push({
+                value: word,
+                index: currentPhraseDef.index
+            });
+            // The second (optional) match is the trailing whitespace.
+            var trailingWhitespace = currentPhraseDef[2];
+            if (trailingWhitespace.length) {
+                segments.push({
+                    value: trailingWhitespace,
+                    index: currentPhraseDef.index + word.length
+                });
+            }
+            currentPhraseDef = regexp.exec(originalString);
+        }
+        // Look for any trailing non-whitespace material after the initial pass.
+        var trailingNonWhitespaceRegexp = /\w+\W+(\w+)$/m;
+        var trailingMatches = originalString.match(trailingNonWhitespaceRegexp);
+        if (trailingMatches) {
+            segments.push({
+                value: trailingMatches[1],
+                index: originalString.length - trailingMatches[1].length
+            });
+        }
+
+        var combinedSegments = [];
+        for (var a = 0; a < segments.length; a++) {
+            for (var b = a + 1; b <= segments.length; b++) {
+                if (a + b > 0) {
+                    var subslice = segments.slice(a, b);
+                    var values = fluid.transform(subslice, function (item) { return item.value; });
+                    var substring = values.join("");
+                    if (substring.match(/\w/)) {
+                        combinedSegments.push({
+                            value: substring,
+                            index: segments[a].index
+                        });
+                    }
+                }
+            }
+        }
+
+        combinedSegments.sort(gpii.diff.sortPhraseDefsByLength);
+        return combinedSegments;
+    }
+};
+
+/**
+ *
+ * Go through a string and create a report that can be used to render a "diff" view.  The result is an array of
+ * segments, which consist of a segment value and details about whether the segment is unchanged, added, or removed
+ * when compared to `leftString`.
+ *
+ * @param `leftString` - The first string, from whose perspective the "diff" results will be represented.
```

Comment:
Types for all parameters for these functions - {String}, {Array}, etc.

Reply:
Done.